service/src/main/java/bio/terra/drshub/services/SignedUrlService.java

```diff
@@ -0,0 +1,54 @@
+package bio.terra.drshub.services;
+
+import bio.terra.bond.model.SaKeyObject;
+import bio.terra.common.iam.BearerToken;
+import bio.terra.drshub.config.DrsHubConfig;
+import bio.terra.drshub.logging.AuditLogEvent;
+import bio.terra.drshub.logging.AuditLogEventType;
+import bio.terra.drshub.logging.AuditLogger;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import java.net.URL;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public record SignedUrlService(
+    DrsHubConfig drsHubConfig,
+    AuthService authService,
+    DrsProviderService drsProviderService,
+    GoogleStorageService googleStorageService,
+    AuditLogger auditLogger) {
+
+  public URL getSignedUrl(
+      String bucket,
+      String objectName,
+      String dataObjectUri,
+      String googleProject,
+      BearerToken bearerToken,
+      String ip) {
+
```

Comment:
I had to spend too much time making sure this is right in terms of business function and how this is expected to work.  I think you did it right in terms of taking the same args as Martha but this really doesn't make sense to me that it takes the `project`, `bucket`, and `objectName`.  That should be enough to uniquely identify an object.  But then it also takes a DRS URI, which seems duplicative at best and contradictory at worst.  For example, what is to stop me from using my DRS URI to authenticate, and then naming some _other_ project/bucket/object combo to try to get access to some _other_ object that isn't identified by the DRS URI?  Best case scenario, that is unintended functionality, worst case the user may be able to access files they weren't intended to have access to.  

Anyway, long story short, I guess making this API to match the same contract as Martha for now makes sense.  But I don't like it.  This just seems… not right?  Or not intuitive?  It can be improved :P

Reply:
We need the drs object uri so that we can get the right fence account from Bond to sign the url. Without it, we have no clue what auth to use!

Reply:
We could do something different here and just provide the compact id or drs hostname. That would definitely work and be less redundant.

Reply:
Looked into this. We could make the bucket and object names optional, and if they don't exist, resolve the drs uri. But, we always need the drs uri, and I don't think its a good idea to encode the splitting of hostnames and compact IDs in the UI.

Reply:
Yeah, I guess to me, Martha can do it all:

`Martha.getResolvedSignedUrl(drsUri)`

This thing can go resolve the object, determine the right provider, get the key from bond, and then sign the url returned by the resolution.  That seems the most obvious to me.  Either that or in any one of the other Martha calls where we resolve the DRS URI and give you back the key from Bond, just also return the signedUrl and leave it up to the client to decide which way they want to use to access the object.  Anyway, I guess I'm bringing us way off topic here, we won't be maintaining this stuff.  But I still have opinions 🙃